moveit_core/robot_model/src/robot_model.cpp

```diff
@@ -1299,10 +1297,15 @@ void moveit::core::RobotModel::setKinematicsAllocators(const std::map<std::strin
         sub_joints.insert(sub->getJointModels().begin(), sub->getJointModels().end());
 
         if (std::includes(joints.begin(), joints.end(), sub_joints.begin(), sub_joints.end()))
-        {
+        {  // sub_joints included in joints: add sub, remove sub_joints from joints set
           std::set<const JointModel*> resultj;
           std::set_difference(joints.begin(), joints.end(), sub_joints.begin(), sub_joints.end(),
                               std::inserter(resultj, resultj.end()));
+          // TODO: instead of maintaining disjoint joint sets here,
+          // should we leave that work to JMG's setSolverAllocators() / computeIKIndexBijection()?
+          // There, a disjoint bijection from joints to solvers is computed anyway.
+          // Underlying question: How do we resolve overlaps? Now the first considered sub group "wins"
+          // But, if the overlap only involves fixed joints, we could consider all sub groups
```

Comment:
Good questions...

MoveIt has been over-engineered in a number of aspects, this being one of them in my opinion...
We don't configure arbitrary new planning groups at runtime and at the moment I believe it is fine to force the user to provide specific ik parameters in ambiguous cases.

Reply:
No, no, no! You misunderstand my point here. I don't want the user to explicitly specify all details of IK for all meta groups. It's excellent that MoveIt can resolve them automatically.
For example, on our bimanual setup, I have planning groups for each arm, fully specifying the IK.
However, the meta group `both_arms`, with subgroups `left_arm` and `right_arm` automatically inherits the IK solvers from their children.

My point here was, that disjointness of joint sets of ensured in two different places.

Reply:
> However, the meta group both_arms, with subgroups left_arm and right_arm automatically inherits the IK solvers from their children.

Of course. But this is no general "overlap", but clear subset.
I didn't read through this code in detail,
but doesn't the code try to tackle the more general overlap case too?
I simply believe that if there is more than one primitive way to decompose a group, then the user should just specify exactly one on their own.
Do you see problems with this (abstract) requirement?

Reply:
The overlap case is handled exactly as you described: the situation is detected and nothing is derived from child groups. However, in that regard, nothing is over-engineered. The existing code is required to detect the situation.

Reply:
OK. To be more precise: The code here, greedily grabs the first sub group which offers an IK solver. A potential conflict with another, later processed group isn't recognized.

The code in `computeIKIndexBijection()` will recognize the overlap and reject IK association (as you requested). However, this is never activated, because the code here only proposed disjoint sets.

As pointed out, we could accept overlaps as long as only fixed joints are overlapping. This is (probably not done yet in `computeIKIndexBijection()`).

To implement the behavior you requested, we need to move disjoint set resolution from `setSolverAllocators()` to `computeIKIndexBijection()` as pointed out in the TODO.
As we are already two of us, agreeing to this behavior, I ask you to file another issue referencing the TODO and this thread - after you merged the present PR ;-)

Reply:
Ok, I looked into the surrounding code a bit more.
I do not agree with the first part of your TODO at the moment.
In my opinion it does not make sense *not* to select the appropriate subset in this part of the code.

`setSolverAllocators`/`computeIKIndexBijection` currently only initializes the solvers individually.
It does not reason about overlap between solver instances at all.
If you feed multiple overlapping solvers in there, they would all be setup "correctly" without complains.
If we implement your proposal, the `JointModelGroup` would be setup with more `SolverAllocatorFn`s than it actually requires and we would "set" only a subset of the arguments.

The second part of the TODO is fully valid though.
Greedy selection might fail in some (at first glance rather strange...) cases even though there could be a correct assignment.

I would propose to change the comment to:

    // TODO: How do we resolve overlaps? Now the first considered sub group "wins"
    // But, if the overlap only involves fixed joints, we could consider all sub groups

Reply:
> If you feed multiple overlapping solvers in there, they would all be setup "correctly" without complains.

This is simply wrong. Did you tried it? I did with a pepper robot, defining arm groups such that they shared the common fixed torso joint. Combining both groups, I only get an interactive marker for one arm...

Reply:
> with a pepper robot, defining arm groups such that they shared the common fixed torso joint. Combining both groups, I only get an interactive marker for one arm

I was not talking about RViz and I'm not sure what happens there. Should it setup one interactive marker for each entry in `group_kinematics_`?
Just for clarification: You removed the logic from `setKinematicsAllocators` that only adds subsets above, and ran RViz?
With this logic intact, this is the behavior I would expect anyway.
I think we should even print a warning if a meta group has a non-empty subset of joints remaining after looking through all "simple" groups as possible subgroups.

I was talking about calls to [`setSolverAllocators`](https://github.com/ros-planning/moveit/blob/kinetic-devel/moveit_core/robot_model/src/joint_model_group.cpp#L616) with the Map argument filled in.
The function just loops through all entries, configures them and registers them in `group_kinematics_`.
[`computeIKIndexBijection`](https://github.com/ros-planning/moveit/blob/kinetic-devel/moveit_core/robot_model/src/joint_model_group.cpp#L575) is called for each solver, but this only sets up the joint_bijection vector *of each solver* and never changes anything in the JMG.
I don't see where an error arises in this part of the code.
This is what I meant by `would all be setup "correctly" without complains`.

Reply:
I opened a new issue #775 to clarify the wrong behavior and continue the discussion on this there. Just for clarification here: I did not touch `setSolverAllocators` at all yet - except the cosmetic changes of this PR.